README.md
# Healthplanet input plugin for Embulk

Retrieve your data from [TANITA's Health Planet website](https://www.healthplanet.jp/) through Health Planet API v1.

This plugin only supports 'innerscan' scope. Other scopes such as 'sphygmomanometer', 'pedometer' and 'smug' are not supported.

## Overview

* **Plugin type**: input
* **Resume supported**: no
* **Cleanup supported**: no
* **Guess supported**: no

## Install

```
$ embulk gem install embulk-input-healthplanet
```

## Configuration

- **login_id**: Login ID for Health Planet website (string, required)
- **password**: Password for Health Planet website (string, required)
- **client_id**: Client ID for your client application using this plugin (string, required)
- **client_secret**: Client Secret for your client application using this plugin (string, required)
- **next_from**: Retrieve data from the time '%Y-%m-%d %H:%M:%S' (string, default: 1 year ago)
- **lang**: Language for column name (string, default: en)
    - en or english: English name such as "weight" or "body fat %"
    - ja or japanese: Japanese name such as "体重" or "体脂肪率"
    - other: as-is API tag such as "6021" or "6022"

## Example

```yaml
in:
  type: healthplanet
  login_id: example_login_id
  password: example_password
  client_id: example.apps.healthplanet.jp
  client_secret: 12345678901123-ExampleClientSecret
  next_from: '2015-01-01 00:00:00'
  lang: ja
exec: {}
out:
  type: file
  path_prefix: ./healthplanet/
  file_ext: csv
  formatter:
    type: csv
    default_timezone: 'Asia/Tokyo'
```

If only new records are required, please use -o option as follows:

```
$ embulk run config.yml -o config.yml
```

## Notice

Health Planet API ver. 1.0 does not export 'Body Water Mass' data. I reported this issue to TANITA. But with no word of thanks, they just replied that they did not fix this issue because this API was free. embulk-input-healthplanet therefore can not export the 'Body Water Mass' data, too.

## References

* [Health Planet API Specification Ver. 1.0 (Japanese)](http://www.healthplanet.jp/apis/api.html)
* [Health Planet からデータをエクスポートするための embulk-input-healthplanet プラグイン - 無印吉澤](http://muziyoshiz.hatenablog.com/entry/2016/01/11/234921 "Health Planet からデータをエクスポートするための embulk-input-healthplanet プラグイン - 無印吉澤")


CHANGELOG.md
## 1.0.1 - 2016-01-17

* [enhancement] Add new config: lang (ja, en, or api)

## 1.0.0 - 2016-01-11

* First release


sample.rb
# -*- coding: utf-8 -*-
require 'faraday'
require 'faraday-cookie_jar'
require 'oga'
require 'nkf'
require 'json'

# Account for Health Planet
LOGIN_ID = ''
PASSWORD = ''

# Credentials for embulk-input-healthplanet, application type "Client Application"
CLIENT_ID = ''
CLIENT_SECRET = ''

# Default redirect URI for Client Application
REDIRECT_URI = 'https://www.healthplanet.jp/success.html'

# Default scope
DEFAULT_SCOPE = 'innerscan'

# Default response type
DEFAULT_RESPONSE_TYPE = 'code'

DEFAULT_GRANT_TYPE = 'authorization_code'

# Setup connection
conn = Faraday.new(:url => 'https://www.healthplanet.jp') do |faraday|
  faraday.request  :url_encoded
  faraday.response :logger
  faraday.use      :cookie_jar
  faraday.adapter  Faraday.default_adapter
end

# API Reference
# http://www.healthplanet.jp/apis/api.html

# Request Authentication page: /oauth/auth
response = conn.get do |req|
  req.url '/oauth/auth'
  req.params[:client_id]     = CLIENT_ID
  req.params[:redirect_uri]  = REDIRECT_URI
  req.params[:scope]         = DEFAULT_SCOPE
  req.params[:response_type] = DEFAULT_RESPONSE_TYPE
end

response = conn.post 'login_oauth.do', { :loginId => LOGIN_ID, :passwd => PASSWORD, :send => '1', :url => "https://www.healthplanet.jp/oauth/auth?client_id=197.YvJT0sFdrb.apps.healthplanet.jp&redirect_uri=https://www.healthplanet.jp/success.html&scope=innerscan,sphygmomanometer,pedometer,smug&response_type=code" }

unless response.status == 302
  p "login failure"
end


# Get /oauth/auth again
response = conn.get do |req|
  req.url '/oauth/auth'
  req.params[:client_id]     = CLIENT_ID
  req.params[:redirect_uri]  = REDIRECT_URI
  req.params[:scope]         = DEFAULT_SCOPE
  req.params[:response_type] = DEFAULT_RESPONSE_TYPE
end

body = NKF.nkf('-Sw', response.body)

# Read oauth_token
document = Oga.parse_html(body)
oauth_token = document.at_xpath('//input[@name="oauth_token"]').get("value")

# Post /oauth/approval.do again
response = conn.post '/oauth/approval.do', { :approval => 'true', :oauth_token => oauth_token }

body = NKF.nkf('-Sw', response.body)

document = Oga.parse_html(body)
code = document.at_xpath('//textarea[@id="code"]').text


# Get request token
response = conn.post do |req|
  req.url '/oauth/token'
  req.params[:client_id]     = CLIENT_ID
  req.params[:client_secret] = CLIENT_SECRET
  req.params[:redirect_uri]  = REDIRECT_URI
  req.params[:code]          = code
  req.params[:grant_type]    = DEFAULT_GRANT_TYPE
end

tokens = JSON.parse(response.body)
access_token = tokens["access_token"]


# Get innerscan data
response = conn.get do |req|
  req.url 'status/innerscan.json'
  req.params[:access_token] = access_token
  # 0: registered time, 1: measured time
  req.params[:date] = 1
  req.params[:from] = '20160101000000'
  req.params[:to]   = '20160201000000'
  req.params[:tag]  = '6021,6022,6023,6024,6025,6026,6027,6028,6029'
end

data = JSON.parse(response.body)

result = {}

data['data'].each do |record|
  result[record['date']] ||= {}
#  result[record['date']]['birth_date']   ||= data['birth_date']
#  result[record['date']]['height']       ||= data['height']
#  result[record['date']]['sex']          ||= data['sex']
  result[record['date']]['model']        ||= record['model']
  result[record['date']][record['tag']]  = record['keydata']
end

p result



sample_mypass.rb
# -*- coding: utf-8 -*-
require 'faraday'
require 'faraday-cookie_jar'
require 'oga'
require 'nkf'
require 'json'

# Account for Health Planet
LOGIN_ID = 'muziyoshiz'
PASSWORD = 'A9kpmdjm'

# Credentials for embulk-input-healthplanet, application type "Client Application"
CLIENT_ID = '197.YvJT0sFdrb.apps.healthplanet.jp'
CLIENT_SECRET = '1452172254933-WxAk9wA3awihlshapceeM2CVu5wyW1PzdA2DlGxE'

# Default redirect URI for Client Application
REDIRECT_URI = 'https://www.healthplanet.jp/success.html'

# Default scope
DEFAULT_SCOPE = 'innerscan'

# Default response type
DEFAULT_RESPONSE_TYPE = 'code'

DEFAULT_GRANT_TYPE = 'authorization_code'

# Setup connection
conn = Faraday.new(:url => 'https://www.healthplanet.jp') do |faraday|
  faraday.request  :url_encoded
  faraday.response :logger
  faraday.use      :cookie_jar
  faraday.adapter  Faraday.default_adapter
end

# API Reference
# http://www.healthplanet.jp/apis/api.html

# Request Authentication page: /oauth/auth
response = conn.get do |req|
  req.url '/oauth/auth'
  req.params[:client_id]     = CLIENT_ID
  req.params[:redirect_uri]  = REDIRECT_URI
  req.params[:scope]         = DEFAULT_SCOPE
  req.params[:response_type] = DEFAULT_RESPONSE_TYPE
end

response = conn.post 'login_oauth.do', { :loginId => LOGIN_ID, :passwd => PASSWORD, :send => '1', :url => "https://www.healthplanet.jp/oauth/auth?client_id=197.YvJT0sFdrb.apps.healthplanet.jp&redirect_uri=https://www.healthplanet.jp/success.html&scope=innerscan,sphygmomanometer,pedometer,smug&response_type=code" }

unless response.status == 302
  p "login failure"
end


# Get /oauth/auth again
response = conn.get do |req|
  req.url '/oauth/auth'
  req.params[:client_id]     = CLIENT_ID
  req.params[:redirect_uri]  = REDIRECT_URI
  req.params[:scope]         = DEFAULT_SCOPE
  req.params[:response_type] = DEFAULT_RESPONSE_TYPE
end

body = NKF.nkf('-Sw', response.body)

# Read oauth_token
document = Oga.parse_html(body)
oauth_token = document.at_xpath('//input[@name="oauth_token"]').get("value")

# Post /oauth/approval.do again
response = conn.post '/oauth/approval.do', { :approval => 'true', :oauth_token => oauth_token }

body = NKF.nkf('-Sw', response.body)

document = Oga.parse_html(body)
code = document.at_xpath('//textarea[@id="code"]').text


# Get request token
response = conn.post do |req|
  req.url '/oauth/token'
  req.params[:client_id]     = CLIENT_ID
  req.params[:client_secret] = CLIENT_SECRET
  req.params[:redirect_uri]  = REDIRECT_URI
  req.params[:code]          = code
  req.params[:grant_type]    = DEFAULT_GRANT_TYPE
end

tokens = JSON.parse(response.body)
access_token = tokens["access_token"]


# Get innerscan data
response = conn.get do |req|
  req.url 'status/innerscan.json'
  req.params[:access_token] = access_token
  # 0: registered time, 1: measured time
  req.params[:date] = 1
  req.params[:from] = '20160101000000'
  req.params[:to]   = '20160201000000'
  req.params[:tag]  = '6021,6022,6023,6024,6025,6026,6027,6028,6029'
end

data = JSON.parse(response.body)

result = {}

data['data'].each do |record|
  result[record['date']] ||= {}
#  result[record['date']]['birth_date']   ||= data['birth_date']
#  result[record['date']]['height']       ||= data['height']
#  result[record['date']]['sex']          ||= data['sex']
  result[record['date']]['model']        ||= record['model']
  result[record['date']][record['tag']]  = record['keydata']
end

p result



lib/embulk/input/healthplanet_api/column.rb
# -*- coding: utf-8 -*-

module Embulk
  module Input
    module HealthplanetApi
      class Column

        def initialize(lang)
          case lang.downcase
          when 'ja', 'japanese'
            @names = {
              :time           => '測定日時',
              :model          => 'モデル',
              :weight         => '体重',
              :body_fat       => '体脂肪率',
              :muscle_mass    => '筋肉量',
              :muscle_score   => '筋肉スコア',
              :visceral_fat2  => '内臓脂肪レベル2',
              :visceral_fat1  => '内臓脂肪レベル1',
              :metabolic_rate => '基礎代謝量',
              :metabolic_age  => '体内年齢',
              :bone_mass      => '推定骨量',
            }
          when 'en', 'english'
            @names = {
              :time           => 'time',
              :model          => 'model',
              :weight         => 'weight',
              :body_fat       => 'body fat %',
              :muscle_mass    => 'muscle mass',
              :muscle_score   => 'muscle score',
              :visceral_fat2  => 'visceral fat level 2',
              :visceral_fat1  => 'visceral fat level 1',
              :metabolic_rate => 'basal metabolic rate',
              :metabolic_age  => 'metabolic age',
              :bone_mass      => 'estimated bone mass',
            }
          else
            # returns as-is API tag
            @names = {
              :time           => 'time',
              :model          => 'model',
              :weight         => '6021',
              :body_fat       => '6022',
              :muscle_mass    => '6023',
              :muscle_score   => '6024',
              :visceral_fat2  => '6025',
              :visceral_fat1  => '6026',
              :metabolic_rate => '6027',
              :metabolic_age  => '6028',
              :bone_mass      => '6029',
            }
          end
        end

        def name(key)
          @names[key]
        end
      end
    end
  end
end


lib/embulk/input/healthplanet.rb
require 'faraday'
require 'faraday-cookie_jar'
require 'oga'
require 'nkf'
require 'json'
require 'time'
require 'embulk/input/healthplanet_api/column'

module Embulk
  module Input

    class Healthplanet < InputPlugin
      Plugin.register_input("healthplanet", self)

      # Default redirect URI for Client Application
      REDIRECT_URI = 'https://www.healthplanet.jp/success.html'

      # Default scope
      DEFAULT_SCOPE = 'innerscan'

      # Default response type
      DEFAULT_RESPONSE_TYPE = 'code'

      # Default grant type
      DEFAULT_GRANT_TYPE = 'authorization_code'

      # All tags for innerscan
      ALL_TAGS = '6021,6022,6023,6024,6025,6026,6027,6028,6029'

      # Health Planet API can response only in 3 months
      RESPONSE_INTERVAL = 60*60*24*30*3

      # embulk-input-healthplanet retrieves data from one year ago by default
      # If you need data more than one year ago, please set 'next_from' parameter
      DEFAULT_FROM_TIME = 60*60*24*365

      def self.transaction(config, &control)
        # configuration code:
        task = {
          # Account for Health Planet
          'login_id' => config.param('login_id', :string),
          'password' => config.param('password', :string),
          # Credential for embulk-input-healthplanet, application type "Client Application"
          'client_id' => config.param('client_id', :string),
          'client_secret' => config.param('client_secret', :string),
          # This plugin retrieves new data after this time
          'next_from' => config.param('next_from', :string, :default => nil)
        }

        lang = config.param('lang', :string, :default => 'en')
        col = HealthplanetApi::Column.new(lang)

        columns = [
          Column.new(0, col.name(:time), :timestamp),
          Column.new(1, col.name(:model), :string),
          Column.new(2, col.name(:weight), :double),
          Column.new(3, col.name(:body_fat), :double),
          Column.new(4, col.name(:muscle_mass), :double),
          Column.new(5, col.name(:muscle_score), :long),
          Column.new(6, col.name(:visceral_fat2), :double),
          Column.new(7, col.name(:visceral_fat1), :long),
          Column.new(8, col.name(:metabolic_rate), :long),
          Column.new(9, col.name(:metabolic_age), :long),
          Column.new(10, col.name(:bone_mass), :double),
          # Not supported by Health Planet API Ver. 1.0
#          Column.new(11, 'body water mass', :string),
        ]

        resume(task, columns, 1, &control)
      end

      def self.resume(task, columns, count, &control)
        task_reports = yield(task, columns, count)

        next_config_diff = task_reports.first
        return next_config_diff
      end

      def init
        login_id = task['login_id']
        password = task['password']
        client_id = task['client_id']
        client_secret = task['client_secret']
        if task['next_from']
          @next_from = Time.strptime(task['next_from'], '%Y-%m-%d %H:%M:%S')
        end

        # Setup connection
        @conn = Faraday.new(:url => 'https://www.healthplanet.jp') do |faraday|
          faraday.request  :url_encoded
          faraday.response :logger
          faraday.use      :cookie_jar
          faraday.adapter  Faraday.default_adapter
        end

        # Request Authentication page: /oauth/auth
        response = @conn.get do |req|
          req.url '/oauth/auth'
          req.params[:client_id]     = client_id
          req.params[:redirect_uri]  = REDIRECT_URI
          req.params[:scope]         = DEFAULT_SCOPE
          req.params[:response_type] = DEFAULT_RESPONSE_TYPE
        end

        # Login and set session information
        response = @conn.post 'login_oauth.do', { :loginId => login_id, :passwd => password, :send => '1', :url => "https://www.healthplanet.jp/oauth/auth?client_id=#{client_id}&redirect_uri=#{REDIRECT_URI}&scope=#{DEFAULT_SCOPE}&response_type=#{DEFAULT_RESPONSE_TYPE}" }

        unless response.status == 302
          # TODO return error in Embulk manner
          print "Login failure\n"
        end

        # Get auth page again with JSESSIONID
        response = @conn.get do |req|
          req.url '/oauth/auth'
          req.params[:client_id]     = client_id
          req.params[:redirect_uri]  = REDIRECT_URI
          req.params[:scope]         = DEFAULT_SCOPE
          req.params[:response_type] = DEFAULT_RESPONSE_TYPE
        end

        # Read oauth_token
        document = Oga.parse_html(NKF.nkf('-Sw', response.body))
        oauth_token = document.at_xpath('//input[@name="oauth_token"]').get('value')

        # Post /oauth/approval.do
        response = @conn.post '/oauth/approval.do', { :approval => 'true', :oauth_token => oauth_token }

        # Read code
        document = Oga.parse_html(NKF.nkf('-Sw', response.body))
        code = document.at_xpath('//textarea[@id="code"]').text

        # Get request token
        response = @conn.post do |req|
          req.url '/oauth/token'
          req.params[:client_id]     = client_id
          req.params[:client_secret] = client_secret
          req.params[:redirect_uri]  = REDIRECT_URI
          req.params[:code]          = code
          req.params[:grant_type]    = DEFAULT_GRANT_TYPE
        end

        tokens = JSON.parse(response.body)
        @access_token = tokens['access_token']
      end

      def run
        from = @next_from.nil? ? (Time.now - DEFAULT_FROM_TIME) : @next_from
        last_date = nil

        while from < Time.now
          to = from + RESPONSE_INTERVAL
          date = innerscan(from, to)
          # Update last_date if any data exists
          last_date = date if date

          # Next request must start from 1 minute later to avoid redundant data
          from = to + 60
        end

        page_builder.finish

        task_report = {}
        unless preview? or last_date.nil?
          # Next request must start from 1 minute later to avoid redundant data
          task_report = { :next_from => (last_date + 60).strftime('%Y-%m-%d %H:%M:%S') }
        end
        return task_report
      end

      def innerscan(from = nil, to = nil)
        response = @conn.get do |req|
          req.url 'status/innerscan.json'
          req.params[:access_token] = @access_token
          # 0: registered time, 1: measured time
          req.params[:date] = 1
          req.params[:from] = from.strftime('%Y%m%d%H%M%S') unless from.nil?
          req.params[:to]   = to.strftime('%Y%m%d%H%M%S')   unless to.nil?
          req.params[:tag]  = ALL_TAGS
        end

        data = JSON.parse(response.body)

        result = {}

        data['data'].each do |record|
          date = Time.strptime(record['date'], '%Y%m%d%H%M')

          result[date] ||= {}
          result[date]['model'] ||= record['model']
          result[date][record['tag']]  = record['keydata']
        end

        dates = result.keys.sort
        last_date = dates.last

        dates.each do |date|
          page = Array.new(11)
          page[0] = date
          result[date].each do |key, value|
            case key
            when 'model'
              page[1] = value
            when '6021'
              # weight
              page[2] = value.to_f
            when '6022'
              # body fat %
              page[3] = value.to_f
            when '6023'
              # muscle mass
              page[4] = value.to_f
            when '6024'
              # muscle score
              page[5] = value.to_i
            when '6025'
              # visceral fat level 2
              page[6] = value.to_f
            when '6026'
              # visceral fat level 1
              page[7] = value.to_i
            when '6027'
              # basal metabolic rate
              page[8] = value.to_i
            when '6028'
              # metabolic age
              page[9] = value.to_i
            when '6029'
              # estimated bone mass
              page[10] = value.to_f
            end
          end

          page_builder.add(page)
        end

        last_date
      end

      def preview?
        begin
          org.embulk.spi.Exec.isPreview()
        rescue java.lang.NullPointerException => e
          false
        end
      end
    end
  end
end
